test/core/libp2p/transport/tcp_integration_test.cpp

```diff
@@ -0,0 +1,256 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <chrono>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <system_error>
+
+#include <gtest/gtest.h>
+#include "libp2p/transport/tcp.hpp"
+#include "testutil/literals.hpp"
+#include "testutil/outcome.hpp"
+
+using namespace libp2p::transport;
+using namespace libp2p::multi;
+using namespace libp2p::connection;
+using std::chrono_literals::operator""s;
+using std::chrono_literals::operator""ms;
+using kagome::common::Buffer;
+
+void logError(const std::error_code &ec) {
+  std::cout << "error(" << ec.value() << "): " << ec.message() << "\n";
+}
+
+void expectConnectionValid(const std::shared_ptr<RawConnection> &conn) {
+  EXPECT_TRUE(conn);
+
+  EXPECT_OUTCOME_TRUE(mar, conn->remoteMultiaddr());
+  EXPECT_OUTCOME_TRUE(mal, conn->localMultiaddr());
+  std::ostringstream s;
+  s << mar.getStringAddress() << " -> " << mal.getStringAddress();
+  std::string tag = s.str();
+  std::cout << tag << '\n';
+}
+
+/**
+ * @given two listeners
+ * @when bound on the same multiaddress
+ * @then get error
+ */
+TEST(TCP, TwoListenersCantBindOnSamePort) {
+  boost::asio::io_context context(1);
+  auto e = context.get_executor();
+  auto transport = std::make_shared<TcpTransport<decltype(e)>>(e);
+
+  auto listener1 = transport->createListener(
+      [](auto &&c) {
+        EXPECT_TRUE(c) << "listener 1 is nullptr";
+        std::cout << "new connection - listener 1\n";
+        return outcome::success();
+      },
+      logError);
+  ASSERT_TRUE(listener1);
+
+  auto listener2 = transport->createListener(
+      [](auto &&c) {
+        EXPECT_TRUE(c) << "listener 2 is nullptr";
+        std::cout << "new connection - listener 2\n";
+        return outcome::success();
+      },
+      logError);
+  ASSERT_TRUE(listener2);
+
+  auto ma = "/ip4/127.0.0.1/tcp/40003"_multiaddr;
+
+  std::cout << "listener 1 starting...\n";
+  ASSERT_TRUE(listener1->listen(ma));
+
+  std::cout << "listener 2 starting...\n";
+  auto r = listener2->listen(ma);
+  ASSERT_EQ(r.error().value(), (int)std::errc::address_in_use);
+
+  using std::chrono_literals::operator""ms;
+  context.run_for(50ms);
```

Comment:
Are there any way to avoid using timers in test? We can end up with problems on CI or some slow machines

Reply:
in this case you would need to do complex test preparation, and gain from this is not that high.

Reply:
We suffered a lot in Iroha because of such tests and spent a lot of time fixing them. If it is not possible to fix them right now, please leave todo with task in Jira to change this in future